cmd/tlssocksproxy/tlssocksproxy.go

```diff
@@ -3,148 +3,159 @@ package main
-
+func serve(ctx context.Context, logger *zap.Logger, localConn net.Conn, remoteAddress string, tlsConfig *tls.Config, connID uint64) {
 	start := time.Now()
 
-	dstConn, errDial := tls.DialWithDialer(&net.Dialer{
-		KeepAlive: -1,
-		Timeout:   defaultTimeout,
-	}, "tcp", destinationAddress, tlsConfig)
+	// Recover if a panic occurs
+	defer recoverAndLogPanic(logger)
+	defer silentClose(localConn)
 
-	if errDial != nil {
-		logger.Warn(
-			"could not reach tls server",
-			zap.Error(errDial),
-		)
+	remoteConn, err := tls.DialWithDialer(&net.Dialer{
+		Timeout: defaultTimeout,
+	}, "tcp", remoteAddress, tlsConfig)
+	if err != nil {
+		logger.Warn("could not reach remote tls server", zap.Error(err))
 		return
 	}
-	defer silentClose(dstConn)
-
-	group, gctx := errgroup.WithContext(ctx)
-
-	group.Go(func() error {
-		srcConn := tlssocks.NewBufferedReader(gctx, srcConn)
-		return copyData("conn->socksConn", dstConn, srcConn)
-	})
-	group.Go(func() error {
-		dstConn := tlssocks.NewBufferedReader(gctx, dstConn)
-		return copyData("socksConn->conn", srcConn, dstConn)
-	})
-
-	if err := group.Wait(); err != nil {
-		switch {
-		case err == io.ErrUnexpectedEOF,
-			err == io.ErrClosedPipe,
-			err == io.EOF,
-			err.Error() == "broken pipe":
-			logger.Warn("Error occurred, while copying data", zap.Error(err))
-		default:
-			logger.Error("Unexpected error occurred while copying the data", zap.Error(err))
-		}
+	defer silentClose(remoteConn)
+
+	p := &proxy{
+		log:   logger,
+		wait:  make(chan struct{}),
+		lconn: localConn,
 	}
 
+	go p.pipe(ctx, remoteConn, localConn)
+	go p.pipe(ctx, localConn, remoteConn)
+
+	<-p.wait
 	logger.Info(
 		"request served",
 		zap.Duration("duration", time.Now().Sub(start)),
+		zap.Uint64("bytes_sent", atomic.LoadUint64(&p.sentBytes)),
+		zap.Uint64("bytes_received", atomic.LoadUint64(&p.receivedBytes)),
+		zap.Uint64("conn_id", connID),
+		zap.String("from", localConn.RemoteAddr().String()),
 	)
 }
 
-func recoverAndLogPanic() {
+type proxy struct {
+	log           *zap.Logger
+	sentBytes     uint64
+	receivedBytes uint64
+	lconn         io.ReadWriteCloser
+	wait          chan struct{}
+	erredMU       sync.Mutex
+	erred         uint32
+}
+
+func (p *proxy) pipe(ctx context.Context, dst io.Writer, src io.Reader) {
+	isLocal := src == p.lconn
+	buff := make([]byte, 65535)
+	for {
+		if ctx.Err() != nil {
+			p.err("context error", ctx.Err())
+			return
+		}
+
+		n, err := src.Read(buff[:])
+		if err != nil {
+			p.err("Read failed", err)
+			return
+		}
+
+		n, err = dst.Write(buff[:n])
+		if err != nil {
+			p.err("Write failed", err)
+			return
+		}
+		if isLocal {
+			atomic.AddUint64(&p.sentBytes, uint64(n))
+		} else {
+			atomic.AddUint64(&p.receivedBytes, uint64(n))
+		}
+	}
+}
+
+func (p *proxy) err(message string, err error) {
+	if atomic.LoadUint32(&p.erred) == 1 {
+		return
+	}
+	if err != io.EOF {
+		p.log.Warn(message, zap.Error(err))
+	}
+
+	p.wait <- struct{}{}
+	atomic.StoreUint32(&p.erred, 1)
```

Comment:
You're not using this value at all. Also, it's always going to be 1, regardless of whether there was an actual error or if the conn was just closed.

Reply:
it gets used somewhere else and should always be one. before it was a boolean and causing a race

Reply:
better would be to check if the channel is closed ;-)